source/common/grpc/async_client_impl.h

```diff
@@ -0,0 +1,215 @@
+#include "common/http/async_client_impl.h"
+#include "common/http/header_map_impl.h"
+#include "common/http/utility.h"
+
+namespace Envoy {
+namespace Grpc {
+
+template <class RequestType, class ResponseType> class AsyncClientStreamImpl;
+
+template <class RequestType, class ResponseType>
+class AsyncClientImpl final : public AsyncClient<RequestType, ResponseType> {
+public:
+  AsyncClientImpl(Upstream::ClusterManager& cm, const std::string& remote_cluster_name)
+      : cm_(cm), remote_cluster_name_(remote_cluster_name) {}
+
+  ~AsyncClientImpl() override {
+    while (!active_streams_.empty()) {
+      active_streams_.front()->cleanup();
+    }
+  }
+
+  // Grpc::AsyncClient
+  AsyncClientStream<RequestType>*
+  start(const google::protobuf::MethodDescriptor& service_method,
+        AsyncClientCallbacks<ResponseType>& callbacks,
+        const Optional<std::chrono::milliseconds>& timeout) override {
+    std::unique_ptr<AsyncClientStreamImpl<RequestType, ResponseType>> grpc_stream{
+        new AsyncClientStreamImpl<RequestType, ResponseType>(*this, callbacks)};
+    Http::AsyncClient::Stream* http_stream =
+        cm_.httpAsyncClientForCluster(remote_cluster_name_)
+            .start(*grpc_stream, Optional<std::chrono::milliseconds>(timeout));
+
+    if (http_stream == nullptr) {
+      callbacks.onRemoteClose(GrpcStatus::Internal);
+      return nullptr;
+    }
+
+    grpc_stream->set_stream(http_stream);
+
+    Http::MessagePtr message = Common::prepareHeaders(
+        remote_cluster_name_, service_method.service()->full_name(), service_method.name());
+    callbacks.onCreateInitialMetadata(message->headers());
+
+    http_stream->sendHeaders(message->headers(), false);
+    grpc_stream->moveIntoList(std::move(grpc_stream), active_streams_);
+
+    return active_streams_.front().get();
+  }
+
+private:
+  Upstream::ClusterManager& cm_;
+  const std::string remote_cluster_name_;
+  std::list<std::unique_ptr<AsyncClientStreamImpl<RequestType, ResponseType>>> active_streams_;
+
+  friend class AsyncClientStreamImpl<RequestType, ResponseType>;
+};
+
+template <class RequestType, class ResponseType>
+class AsyncClientStreamImpl : public AsyncClientStream<RequestType>,
+                              Http::AsyncClient::StreamCallbacks,
+                              LinkedObject<AsyncClientStreamImpl<RequestType, ResponseType>> {
+public:
+  AsyncClientStreamImpl(AsyncClientImpl<RequestType, ResponseType>& parent,
+                        AsyncClientCallbacks<ResponseType>& callbacks)
+      : parent_(parent), callbacks_(callbacks) {}
+
+  // Http::AsyncClient::StreamCallbacks
+  void onHeaders(Http::HeaderMapPtr&& headers, bool end_stream) override {
+    if (remote_closed_) {
+      return;
+    }
+    if (Http::Utility::getResponseStatus(*headers) != enumToInt(Http::Code::OK)) {
+      reset();
+      callbacks_.onRemoteClose(GrpcStatus::Internal);
+      return;
+    }
+    if (end_stream) {
+      onTrailers(std::move(headers));
```

Comment:
http://www.grpc.io/docs/guides/wire.html refers to this as a Trailers-Only response.